DigitalLearningSolutions.Web/Scripts/trackingSystem/centreCourseSetup.ts

```diff
@@ -0,0 +1,54 @@
+const copyCourseLinkClass = 'copy-course-link';
+const copyLinkIdPrefix = 'copy-course-';
+const launchCourseButtonIdPrefix = 'launch-course-';
+
+setupCourseLinkClipboardCopiers();
+
+function setupCourseLinkClipboardCopiers() {
+  const copyCourseLinks = document.getElementsByClassName(copyCourseLinkClass);
+
+  for (let i = 0; i < copyCourseLinks.length; i += 1) {
+    const currentLink = copyCourseLinks[i];
+    const linkId = currentLink.id;
+    const customisationId = linkId.slice(copyLinkIdPrefix.length);
+
+    currentLink.addEventListener('click', (e) => {
+      e.preventDefault();
+      e.stopPropagation();
+      copyLaunchCourseLinkToClipboard(customisationId);
+    });
+  }
+}
+
+function copyLaunchCourseLinkToClipboard(customisationId: string) {
+  const launchCourseButtonId = launchCourseButtonIdPrefix + customisationId;
+  const copyText = document.getElementById(launchCourseButtonId) as HTMLAnchorElement;
+  const link = copyText.href;
+  const succeeded = copyTextToClipboard(link);
+  let alertMessage: string;
+
+  if (succeeded) {
+    alertMessage = `Copied the text: ${link}`;
+  } else {
+    alertMessage = `Copy not supported or blocked. Try manually selecting and copying: ${link}`;
+  }
+
+  alert(alertMessage);
+}
+
+function copyTextToClipboard(textToCopy: string) : boolean {
+  const hiddenInput = document.body.appendChild(document.createElement('input'));

```

Comment:
Could we just use launch course button instead of creating and removing a new element?

Reply:
Unfortunately it seems that for document.ExecCommand("copy") to work correctly, it needs an element that could have the text to copy selectable/highlightable. That's why an input or textarea seems to be used typically.

I did do a bit of further research though, and it looks like navigator.clipboard.writeText works on all browsers except IE, so I've moved this method to be a fallback. Mostly stolen from this https://stackoverflow.com/a/30810322